compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/body/resolve/FirDeclarationsResolveTransformer.kt

```diff
@@ -724,6 +727,20 @@ open class FirDeclarationsResolveTransformer(transformer: FirBodyResolveTransfor
         anonymousFunction: FirAnonymousFunction,
         data: ResolutionMode
     ): CompositeTransformResult<FirStatement> {
+        if (anonymousFunction.returnTypeRef is FirImplicitTypeRef) {
```

Comment:
This part looks hacky. I'd propose this case should be fixed somewhere inside lambda resolve. cc @demiurg906

Reply:
I agree. Nothing should be applied to lambda in `transformAnonymousFunction` before `when` on resolution mode (because we visit lambdas which are arguments of other calls twice with this function)

Actually there is a proper place for it already in `transformAnonymousFunctionWithLambdaResolution`, it just should be changed a little bit

Reply:
It seems `transformAnonymousFunctionWithLambdaResolution` is too late, since we already resolve the lambda with an incorrect expected type. We can replace/transform return type of the lambda there, but the use site of lambda assumes the return type of the lambda match with expected parameter type, which shouldn't be the case.

Rather, I found `extractLambdaInfoFromFunctionalType` is a better fit to _enforce_ the return type of a function literal without body.